sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala

```diff
@@ -109,7 +114,10 @@ trait CodegenSupport extends SparkPlan {
     * Consume the columns generated from current SparkPlan, call it's parent.
     */
   final def consume(ctx: CodegenContext, input: Seq[ExprCode], row: String = null): String = {
-    if (input != null) {
+    // We check if input expressions has same length as output when:
+    // 1. parent can't consume UnsafeRow and input is not null.
+    // 2. parent consumes UnsafeRow and row is null.
+    if ((input != null && !parent.consumeUnsafeRow) || (parent.consumeUnsafeRow && row == null)) {
```

Comment:
When the child knows its parent can consume UnsafeRow, it can choose to pass an UnsafeRow and empty input. If so, we don't need to check if `input.length == output.length` here.